### ISSUE TITLE
Access full type name?

### ISSUE THREAD

**Issue by bejar37** [krzysztofzablocki/Sourcery#139]
I have source code like this:

```swift
enum Blah { //Blah used as namespace
    struct Bar {
        let value: Int
    }
    struct Foo {
        let bar: Bar
    }
}
```

I want to write a template that generates code for `Foo` like this:

```swift
extension Blah.Foo {
    func someFunc() {
       let b: Blah.Bar = ... // I need the fully-qualified name of `Bar` here
    }
}
```

Right now, I can't access `bar`'s full type name, only the local one. This makes it so that I have to write this definition for `Foo` in order for my template to work:

```swift
...
struct Foo {
    let bar: Blah.Bar
}
...
```

I looked and it doesn't seem like there's a way to not have to use the fully-qualified `Blah.Bar` typename in the `Foo` declaration. Is that true?

Hope this makes sense!

**Comment by bejar37** [comment]
A stored variable's type is `nil` a lot of times, I'm not too sure why. In my project for example properties of `Bool` and `[String]` have `nil` types, but properties of `String` do have the correct type. I wasn't sure if this behavior was known as a limitation.

**Comment by ilyapuchka** [comment]
Yes, this is a know limitation util we start to scan standard library (#87). `type` is set for properties of types, defined or extended somehow in scanned target. Bool is not defined in the scanned target. But you can always access type name with `typeName`property.
So if you want to have `type` for bool properties you can extend it, but it will only contain information from this extension, nothing from standard library.
But this limitation should not effect full type name. Can you paste a piece of code for which Sourcery fails to return proper type name?

Ilya Puchka

Отправлено с iPhone

**Comment by bejar37** [comment]
That's interesting, it seems like `String` variables have a defined type properties. Is that because I have `extension String` in my codebase?

I don't think I can get the proper type name here:

```swift
enum Blah {
    struct Foo {
        let i: Int
    }
    struct Bar {
        let foos: [Foo]
    }
}
```
How do I get `Blah.Foo` from `foos`? Right now `type` is `nil`, and `typeName` is `[Foo]`, which won't compile in an extension.

**Comment by ilyapuchka** [comment]
Yeah, I guess in this case we do not infer the type correctly, as we don't parse the name of the type `[Foo]` when resolving it. But we do that when we try to infer type from initial value, so we can improve that for that case too. Until then I guess using the fully qualified type name in your code is the only way (you can use typealias to make it nicer)

**Comment by bejar37** [comment]
Does it make sense to add this to the 1.0 milestone @krzysztofzablocki? The workaround is not really onerous, but seems like we want to support this eventually.

**Comment by krzysztofzablocki** [comment]
@bejar37 can you create a simple test[s] that fails that we can run against? that would be very helpful in getting this figured out in sensible way

**Comment by ilyapuchka** [comment]
I've started to work on the fix for that already, will also fix few other issues I've discovered.

Ilya Puchka

Отправлено с iPhone

**Comment by ilyapuchka** [comment]
@krzysztofzablocki I'm going to approach this issue by changing what type names do we store for variables. Currently we only store name as it is used in source code and we attempt to resolve it to actual type if this name is type alias. But we don't do anything if this name is a name of contained type, or wrapped in array or dictionary literal. That results in this issue in case contained type is used with not fully qualified name and actual type is not set.
I think if we will store fully qualified names we will not cause any issues in templates as type names accessed either with `typeName` or `type.name` will be always complete. Looks like `actualTypeName` that for now was only used for type aliases is a good place to store fully qualified name.

I was able to fix particular case that @bejar37 reported, but then started to apply it more broadly and I think I got a bit lost, so will probably start over. I also think it might make sense not to try to fix all possible cases right away, as there are a lot of them (type aliases, tuples, method parameters with attributes).
Does it sound reasonable?
